File: typescript/src/guard.ts
import { PolicyViolationError } from "./errors";
import { LicenseValidator } from "./license";
import type { CheckResult, GuardOptions } from "./types";

export class Guard {
  private enforcementMode: "realtime" | "batch" | "advisory";
  private rateLimit: number;
  private readonly validator: { validate: () => Promise<any> };
  private readonly initPromise: Promise<void>;

  constructor(options: GuardOptions) {
    this.enforcementMode = options.enforcementMode ?? "realtime";
    this.rateLimit = options.rateLimit ?? 500;
    this.validator =
      options.validator ??
      new LicenseValidator(
        options.apiKey,
        options.licenseServerUrl ?? LicenseValidator.DEFAULT_URL,
        options.timeoutMs ?? 800,
      );

    this.initPromise = this.initialize();
  }

  private async initialize(): Promise<void> {
    const license = await this.validator.validate();
    this.enforcementMode = (license.enforcement_mode as "realtime" | "batch" | "advisory") ?? this.enforcementMode;
    this.rateLimit = license.rate_limit ?? this.rateLimit;
  }

  async check(prompt: string, model: string, userId: string): Promise<CheckResult> {
    await this.initPromise;
    return this.evaluate(prompt);
  }

  async checkOutput(output: string, model: string, userId?: string): Promise<CheckResult> {
    await this.initPromise;
    return this.evaluate(output);
  }

  private evaluate(content: string): CheckResult {
    const start = performance.now();
    const blocked = content.includes("<struxel:block>");
    const latency = performance.now() - start;

    const result: CheckResult = {
      blocked,
      reason: blocked ? "Content violates policy" : null,
      policy_id: blocked ? "policy.blocked_content" : null,
      enforcement_mode: this.enforcementMode,
      latency_ms: latency,
    };

    if (!blocked) {
      return result;
    }

    if (this.enforcementMode === "realtime") {
      throw new PolicyViolationError("Content blocked by Struxel Guard.", result);
    }

    if (this.enforcementMode === "batch") {
      console.info("Struxel violation recorded", result.reason);
    } else {
      console.warn("Struxel advisory warning", result.reason);
    }

    return result;
  }
}

File: go/guard.go
package struxelsdk

import (
	"log"
	"strings"
	"time"
)

type Guard struct {
	enforcementMode string
	rateLimit       int
	validator       *LicenseValidator
}

func NewGuard(apiKey, enforcementMode string, rateLimit int, licenseServerURL string, timeoutMS int) (*Guard, error) {
	if enforcementMode == "" {
		enforcementMode = "realtime"
	}
	if rateLimit == 0 {
		rateLimit = 500
	}
	if timeoutMS == 0 {
		timeoutMS = 800
	}

	validator := NewLicenseValidator(apiKey, licenseServerURL, timeoutMS)
	license, err := validator.Validate()
	if err != nil {
		return nil, err
	}

	if license.EnforcementMode != "" {
		enforcementMode = license.EnforcementMode
	}
	if license.RateLimit != 0 {
		rateLimit = license.RateLimit
	}

	return &Guard{enforcementMode: enforcementMode, rateLimit: rateLimit, validator: validator}, nil
}

func (g *Guard) Check(prompt, model, userID string) (CheckResult, error) {
	return g.evaluate(prompt)
}

func (g *Guard) CheckOutput(output, model string, userID ...string) (CheckResult, error) {
	return g.evaluate(output)
}

func (g *Guard) evaluate(content string) (CheckResult, error) {
	start := time.Now()
	blocked := strings.Contains(content, "<struxel:block>")
	latencyMS := float64(time.Since(start).Nanoseconds()) / 1_000_000

	var reason *string
	var policyID *string
	if blocked {
		r := "Content violates policy"
		p := "policy.blocked_content"
		reason = &r
		policyID = &p
	}

	result := CheckResult{
		Blocked:         blocked,
		Reason:          reason,
		PolicyID:        policyID,
		EnforcementMode: g.enforcementMode,
		LatencyMS:       latencyMS,
	}

	if !blocked {
		return result, nil
	}

	if g.enforcementMode == "realtime" {
		return result, &PolicyViolationError{Message: "Content blocked by Struxel Guard.", Result: result}
	}

	if g.enforcementMode == "batch" {
		log.Printf("Struxel violation recorded: %s", *reason)
	} else {
		log.Printf("Struxel advisory warning: %s", *reason)
	}

	return result, nil
}
